airflow/models/param.py

```diff
@@ -0,0 +1,208 @@
+            If true and validations fails, the return value would be None.
+        :type suppress_exception: bool
+        """
+        try:
+            final_val = value or self.default
+            jsonschema.validate(final_val, self.schema, format_checker=FormatChecker())
+            self.default = final_val
+        except ValidationError as err:
+            if suppress_exception:
+                return None
+            raise ValueError(err) from None
+        return final_val
+
+    def dump(self) -> dict:
+        """Dump the Param as a dictionary"""
+        out_dict = {'__class': f'{self.__module__}.{self.__class__.__name__}'}
+        out_dict.update(self.__dict__)
+        return out_dict
+
+
+class ParamsDict(dict):
+    """
+    Class to hold all params for dags or tasks. All the keys are strictly string and values
+    are converted into Param's object if they are not already. This class is to replace param's
+    dictionary implicitly and ideally not needed to be used directly.
+    """
+
+    def __init__(self, dict_obj: Optional[Dict] = None, suppress_exception: bool = False):
+        """
+        Init override for ParamsDict
+        :param dict_obj: A dict or dict like object to init ParamsDict
+        :type dict_obj: Optional[dict]
+        :param suppress_exception: Flag to suppress value exceptions while initializing the ParamsDict
+        :type suppress_exception: bool
+        """
+        params_dict = {}
+        dict_obj = dict_obj or {}
+        for k, v in dict_obj.items():
+            if not isinstance(v, Param):
+                params_dict[k] = Param(v)
+            else:
+                params_dict[k] = v
+        dict.__init__(self, params_dict)
+        self.suppress_exception = suppress_exception
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        """
+        Override for dictionary's ``setitem`` method. This method make sure that all values are of
+        Param's type only.
+
+        :param key: A key which needs to be inserted or updated in the dict
+        :type key: str
+        :param value: A value which needs to be set against the key. It could be of any
+            type but will be converted and stored as a Param object eventually.
+        :type value: Any
+        """
+        try:
+            param = dict.__getitem__(self, key)  # check if the param is in dict already
+            # if the new value is of Param type, then just use it otherwise call resolve on it
+            if isinstance(value, Param):
+                param = value
+            else:
+                param.resolve(value=value, suppress_exception=self.suppress_exception)
+        except KeyError:
+            # if the key isn't there already and if the value is of Param type,
+            # then use it otherwise create a new Param object
+            param = value if isinstance(value, Param) else Param(value)
```

Comment:
Slightly simpler version (that doesn't needlessly get the param if we don't need to check it etc.)

```suggestion
        if isinstance(value, Param):
            param = value
        elif key in self:
            param = dict.__getitem__(self, key)
            param.resolve(value=value, suppress_exception=self.suppress_exception)
        else:
            # if the key isn't there already and if the value isn't of Param type create a new Param object
            param = Param(value)
```